src/rebar_erlc_compiler.erl

```diff
@@ -315,77 +234,75 @@ update_erlcinfo(G, Source, Dirs) ->
+                    case Modified orelse MaxModified > LastUpdated of
+                        true -> modified;
+                        false -> unmodified
+                    end
             end;
         false ->
-            modify_erlcinfo(G, Source, Dirs),
-            modified
+            modify_erlcinfo(G, Source, filelib:last_modified(Source), filename:dirname(Source), Dirs)
+    end.
+
+update_erlcinfo_fun(G, Dirs) ->
+    fun(Erl, Modified) ->
+        case update_erlcinfo(G, Dirs, Erl) of
+            modified -> true;
+            unmodified -> Modified
+        end
     end.
 
-modify_erlcinfo(G, Source, Dirs) ->
+update_max_modified_deps(G, Source) ->
+    MaxModified = lists:max(lists:map(
+        fun(File) -> {_, MaxModified} = digraph:vertex(G, File), MaxModified end,
+        [Source|digraph:out_neighbours(G, Source)])),
+    digraph:add_vertex(G, Source, MaxModified),
+    MaxModified.
+
+modify_erlcinfo(G, Source, LastModified, Dir, Dirs) ->
     {ok, Fd} = file:open(Source, [read]),
-    Incls = parse_attrs(Fd, []),
+    Incls = parse_attrs(Fd, [], Dir),
     AbsIncls = expand_file_names(Incls, Dirs),
     ok = file:close(Fd),
-    LastUpdated = {date(), time()},
-    digraph:add_vertex(G, Source, LastUpdated),
+    digraph:add_vertex(G, Source, LastModified),
+    digraph:del_edges(G, digraph:out_edges(G, Source)),
     lists:foreach(
       fun(Incl) ->
-              update_erlcinfo(G, Incl, Dirs),
+              update_erlcinfo(G, Dirs, Incl),
               digraph:add_edge(G, Source, Incl)
-      end, AbsIncls).
+      end, AbsIncls),
+    modified.
 
-restore_erlcinfo(Config) ->
-    File = erlcinfo_file(Config),
-    G = digraph:new(),
-    case file:read_file(File) of
+restore_erlcinfo(G, InclDirs) ->
+    case file:read_file(erlcinfo_file()) of
         {ok, Data} ->
-            try binary_to_term(Data) of
-                Erlcinfo ->
-                    ok = check_erlcinfo(Config, Erlcinfo),
-                    #erlcinfo{info=ErlcInfo} = Erlcinfo,
-                    {Vs, Es} = ErlcInfo,
-                    lists:foreach(
-                      fun({V, LastUpdated}) ->
-                              digraph:add_vertex(G, V, LastUpdated)
-                      end, Vs),
-                    lists:foreach(
-                      fun({V1, V2}) ->
-                              digraph:add_edge(G, V1, V2)
-                      end, Es)
-            catch
-                error:badarg ->
-                    ?ERROR(
-                       "Failed (binary_to_term) to restore rebar info file."
-                       " Discard file.", []),
-                    ok
-            end;
-        _Err ->
+            % Since externally passed InclDirs can influence erlcinfo graph (see
+            % modify_erlcinfo), we have to check here that they didn't change.
+            #erlcinfo{vsn=?ERLCINFO_VSN, info={Vs, Es, InclDirs}} =
+                binary_to_term(Data),
```

Comment:
No, because any errors are catched when calling `restore_erlcinfo`. It should be only ever called at exactly one place so catching error outside the function should be ok.